bundles/org.openhab.binding.webthing/src/main/java/org/openhab/binding/webthing/internal/client/WebSocketConnectionImpl.java

```diff
@@ -0,0 +1,169 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.webthing.internal.client;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+import org.eclipse.jetty.websocket.api.Session;
+import org.eclipse.jetty.websocket.api.WebSocketListener;
+import org.eclipse.jetty.websocket.api.WebSocketPingPongListener;
+import org.eclipse.jetty.websocket.api.annotations.*;
+import org.jetbrains.annotations.NotNull;
+import org.openhab.binding.webthing.internal.client.dto.PropertyStatusMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
+
+/**
+ * The WebsocketConnection implementation
+ *
+ * @author Gregor Roth - Initial contribution
+ */
+public class WebSocketConnectionImpl implements WebSocketConnection, WebSocketListener, WebSocketPingPongListener {
+    private static final BiConsumer<String, Object> EMPTY_PROPERTY_CHANGED_LISTENER = (String propertyName,
+            Object value) -> {
+    };
+    private final Logger logger = LoggerFactory.getLogger(WebSocketConnectionImpl.class);
+    private final Duration pingPeriod;
+    private final Consumer<String> errorHandler;
+    private final ScheduledFuture<?> watchDogHandle;
+    private final ScheduledFuture<?> pingHandle;
+    private final Map<String, BiConsumer<String, Object>> propertyChangedListeners = new HashMap<>();
+    private final AtomicReference<Instant> lastTimeReceived = new AtomicReference<>(Instant.now());
+    private final AtomicReference<Session> sessionRef = new AtomicReference<>(null);
+
+    /**
+     * constructor
+     *
+     * @param executor the executor to use
+     * @param errorHandler the errorHandler
+     * @param pingPeriod the period pings should be sent
+     */
+    WebSocketConnectionImpl(ScheduledExecutorService executor, Consumer<String> errorHandler, Duration pingPeriod) {
+        this.errorHandler = errorHandler;
+        this.pingPeriod = pingPeriod;
+        this.pingHandle = executor.scheduleAtFixedRate(this::sendPing, pingPeriod.dividedBy(2).toMillis(),
+                pingPeriod.toMillis(), TimeUnit.MILLISECONDS);
```

Comment:
A ping doesn't seem to be too timing sensitive. Can this be `scheduleWithFixedDelay`?

Reply:
yes